### ISSUE TITLE
Add site wide search as an option

### ISSUE THREAD

**Issue by h-enk** [h-enk/doks#630]
## Summary

Brief explanation of the feature.

- Include option to include all sections
- Include option to define sections to include

### Basic example

Include a basic example or links here.

### Motivation

Why are we doing this? What use cases does it support? What is the expected outcome?

https://github.com/h-enk/doks/discussions/395

**Comment by james-d-elliott** [comment]
If you're interested, I implemented something similar in this commit: https://github.com/authelia/website/commit/a902b3335b62c089eb7954aef04ee739c96d2e17

Would this roughly suffice if I included the option to enable it for all sections? I'm relatively sure that part is easy.

**Comment by h-enk** [comment]
Interesting! I'll be looking into this and get back to you.

**Comment by h-enk** [comment]
James, thanks for all your input and work on Doks. I really, really appreciate it!

I took a closer look and I thinks it's OK to proceed from here with a PR. Are you willing to do so or are you too busy?

I think the default should be to show the search box on all pages, and to make only the docs section searchable by default.

BTW, inspired by your setup, I used `sections.sectionNav` for [configuring section navigation](https://github.com/h-enk/doks/blob/master/config/_default/params.toml#L78-L79).

P.S. You OK with me adding [Authelia](https://www.authelia.com/) to the Doks showcase?

**Comment by james-d-elliott** [comment]
> P.S. You OK with me adding [Authelia](https://www.authelia.com/) to the Doks showcase?

Yep that's fine by me!


> I took a closer look and I thinks it's OK to proceed from here with a PR. Are you willing to do so or are you too busy?

I think I have time to do this, it's a pretty small feature and I think it's already implemented on the Authelia site; albeit hacked together. 

What I'll do is I'll refine the way we do it and make it more consistent with the change you mentioned, make the PR and if I don't have time to finish it off for whatever reason you can take it over locally or commit directly to my branch (I always enable that option). 

> I think the default should be to show the search box on all pages, and to make only the docs section searchable by default.

I think I understand. I think option 2 works the best for both, what do you think?

### Showing the search box by default on all pages

We will benefit from adding a list of sections for this functionality. But we have options for how we control if it is only shown on some sections.

1. Enable the searchbox in all sections if the list of sections length is 0. Drawback is they can't disable it entirely maybe?
2. Boolean which disables showing it on all pages if `true` and instead relies on the list of sections.

### Searchable sections

Likewise there is benefit in a list of sections for this functionality. But we have options for how we control if a section is searchable:

1. Enable searching all sections if the list of sections length is 0.
2. Boolean which enables the search cache for all sections if `true`.

**Comment by h-enk** [comment]
Great that you're picking this up. And, totally understand if I've to take over /step in at some point.

Note, that we also have the option `flexSearch = true`, which disables the search functionality completely.

I think we could do with three options to make this work (or, am I missing something, think cache?)

| flexSearch | searchShow | searchableSections | Show search box | Searchable sections |
|---|---|---|---|---|
| FALSE | N/R | N/R | On none pages  | N/R |
| TRUE | EMPTY | EMPTY | On all pages | Docs only |
| TRUE | Defined | EMPTY | On section pages defined | Docs only |
| TRUE | Defined | Defined | On section pages defined | Sections defined |
| TRUE | ALL | ALL | On all section pages | All sections |

**Comment by james-d-elliott** [comment]
The cache is what I meant by searchable sections, flexsearch has data which is used to perform the search, which we populate for lack of a better term with info from specific sections.

WHat do you mean by N/R? it's probably obvious but not jumping out to me.

**Comment by h-enk** [comment]
> The cache is what I meant by searchable sections, flexsearch has data which is used to perform the search, which we populate for lack of a better term with info from specific sections.

Ah, ok /clear

> WHat do you mean by N/R? it's probably obvious but not jumping out to me. Maybe Not Relevant?

Yes, indeed. Sorry about not being clear

**Comment by james-d-elliott** [comment]
One thing missing from the equation is the changes to the show search box will not be intuitive disable/enable for the frontpage since the .Section is named ``. Maybe we psuedo-name it for these purposes as `DoksIndex`.

Currently my snippet for determining if it should show is this:

```html
        {{ if and .Site.Params.options.flexSearch (or (eq (len .Site.Params.options.searchSectionsShow) 0) (in .Site.Params.options.searchSectionsShow .Section)) -}}
        <hr class="text-black-50 my-4 d-lg-none">
        <form class="doks-search position-relative flex-grow-1 ms-lg-auto me-lg-2">
          <input id="search" class="form-control is-search" type="search" placeholder="Search docs..." aria-label="Search docs..." autocomplete="off">
          <div id="suggestions" class="shadow bg-white rounded d-none"></div>
        </form>
        {{ end -}}
```

I'm suggesting it should be this:

```html
        {{ if and .Site.Params.options.flexSearch (or (eq (len .Site.Params.options.searchSectionsShow) 0) (in .Site.Params.options.searchSectionsShow .Section) (and (eq .Section "") (in .Site.Params.options.searchSectionsShow "DoksIndex"))) -}}
        <hr class="text-black-50 my-4 d-lg-none">
        <form class="doks-search position-relative flex-grow-1 ms-lg-auto me-lg-2">
          <input id="search" class="form-control is-search" type="search" placeholder="Search docs..." aria-label="Search docs..." autocomplete="off">
          <div id="suggestions" class="shadow bg-white rounded d-none"></div>
        </form>
        {{ end -}}
```

**Comment by h-enk** [comment]
Nice work!

Yes, next to sections we have pages off course. Do you think it would be possible (and relatively easy) to accommodate the logic for page slugs — like "home", "contact", and "privacy-policy" for instance?

I tested + verified the cases from the table. I only couldn't trigger the last one?

**Comment by james-d-elliott** [comment]
Did you try with:

```
  searchSectionsShow = "ALL"
  searchSectionsIndex = "ALL"
```

This shows for all pages to me, but maybe I'm missing something. Here's a patch to show what I tried:

```diff
Index: content/en/privacy-policy/index.md
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/content/en/privacy-policy/index.md b/content/en/privacy-policy/index.md
--- a/content/en/privacy-policy/index.md	(revision b532c69bd769353490e412e22d790e484aeb91dc)
+++ b/content/en/privacy-policy/index.md	(date 1656152711012)
@@ -3,7 +3,7 @@
 description: "We do not use cookies and we do not collect any personal data."
 date: 2020-08-27T19:23:18+02:00
 lastmod: 2020-08-27T19:23:18+02:00
-draft: true
+draft: false
 images: []
 ---
 
Index: config/_default/params.toml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/config/_default/params.toml b/config/_default/params.toml
--- a/config/_default/params.toml	(revision b532c69bd769353490e412e22d790e484aeb91dc)
+++ b/config/_default/params.toml	(date 1656152541948)
@@ -83,8 +83,8 @@
   clipBoard = true
   instantPage = true
   flexSearch = true
-  searchSectionsShow = []
-  searchSectionsIndex = []
+  searchSectionsShow = "ALL"
+  searchSectionsIndex = "ALL"
   darkMode = true
   bootStrapJs = true
   breadCrumb = false
Index: content/en/contact/index.md
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/content/en/contact/index.md b/content/en/contact/index.md
--- a/content/en/contact/index.md	(revision b532c69bd769353490e412e22d790e484aeb91dc)
+++ b/content/en/contact/index.md	(date 1656152746128)
@@ -3,7 +3,7 @@
 description: "Drop us an email."
 date: 2020-08-27T19:25:12+02:00
 lastmod: 2020-08-27T19:25:12+02:00
-draft: true
+draft: false
 images: []
 ---
```

**Comment by james-d-elliott** [comment]
I think what maybe is going on is what you expected to configure them as wasn't how I understood it and maybe we just need to figure that out.

**Comment by h-enk** [comment]
I got it now, it was my bad — I just had to rebuild the site manually.

**Comment by h-enk** [comment]
> Maybe we psuedo-name it for these purposes as `DoksIndex`.

That's fine. Or, is `HomePage` clearer?

**Comment by h-enk** [comment]
I really like this functionality — very powerful + brilliant

**Comment by james-d-elliott** [comment]
I'm really glad. Also HomePage works for me. I'm not quite sure how to do that with the indexing (showing it is easy) but I'll figure it out.

**Comment by james-d-elliott** [comment]
> 

Oh I'm betting it's because changing a slice into a string or vice versa. I'm guessing it has some kind of issue with the type change. I couldn't think of a cleaner way to do this, allowing an `ALL` slice element seems like a lot harder to maintain, but we could technically do it (just check if it's in the slice).

**Comment by james-d-elliott** [comment]
I think that should work, I tested it and it looks to work fine. Was much easier than I thought!

I think this covers all the bases. I made most things individual commits so we can pick and choose if we want certain things a little. We'll want to squash it though probably. I think all that's left is decisions on if the naming convention and position of those params is correct (let me know if you identify anything else).
